corehq/apps/export/tasks.py

```diff
@@ -81,14 +81,9 @@ def populate_export_download_task(export_instances, filters, download_id, filena
 
 @task(queue=SAVED_EXPORTS_QUEUE, ignore_result=True)
 def _start_export_task(export_instance_id, last_access_cutoff):
-    keys = ['rebuild_export_task_%s' % export_instance_id]
-    timeout = 48 * 3600  # long enough to make sure this doesn't get called while another one is running
-    with CriticalSection(keys, timeout=timeout, block=False) as locked_section:
```

Comment:
A little hard to wrap my head around this, but it seems like you moved a locking/concurrency check from inside the task to the caller of the task, yeah? Just thinking out loud I could imagine that could cause problems in rare scenarios - e.g. you check the task status from two threads at the same time, both see there is nothing there, then both start the task at the same instant resulting in the check not working and the task running in parallel. Realize it's a bit of an edge case, but seems less safe than doing it in the task. Granted, I don't really understand why this was in a `CriticalSection` in the first place or what problems arise if that ever trips. 

Does this commit fix a bug? Or just result in fewer noops in the queue? Just wondering if we can do the check outside the task but keep the `CriticalSection` logic in place for edge cases.

Reply:
Saved exports have a single place for a payload, so if two tasks run at the same time, one will finish, be saved, and then the second one will finish and overwrite the result of the first task.  These are non-trivially expensive to compute, which I believe is why this is locked.

That was added in https://github.com/dimagi/commcare-hq/pull/17848.  (@snopoke you may be interested in this change.)

The impetus for this last commit is that this progress bar operates on the task itself.  If a task is queued, it then becomes _the_ task for that export instance, and we lose the pointer to any previous task.  Basically, if a task is in progress, and another one is queued, it will noop, and the user will see the (senseless) progress bar from the second task.  To prevent this, we need to know whether the task will conflict before calling it.

I also considered just checking the lock before calling the task, but that wouldn't cover tasks that are queued but not yet started.  It wouldn't be difficult to add some locking back in, for instance in the code that queues a task, to be sure that's never executed in parallel.  No strong feelings there.  What do you think?

Reply:
Thanks for the explanation. It sounds like the failure mode is basically "do some work twice" which, under the extremely rare edge case I pointed out, seems totally fine. So I think your change makes sense and no need for it to get any more complicated.